### PR TITLE
Pagination for getting claims by targetId

### PR DESCRIPTION
**Pull request by bogdan-zaharia**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/DOC-1184

### *What changes have we introduced*

Added pagination for the endpoint to retrieve claims by targetId.

- Added gateway function that returns paginated claims by targetId
- Added new use case to get claims by targetId
- Added helper functions to encode and decode to/from base64url
- Added new request and response objects for the endpoint to get claims by targetId
- Added unit tests and end to end tests

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [ ] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (andrewclowes) [inline .vscode/launch.json:1]: Do we want to include this VSCode specific file? Not everybody uses this editor for .NET on the team.

File: DocumentsApi/V1/Validators/PaginatedClaimRequestValidator.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        public PaginatedClaimRequestValidator()
+        {
+            RuleFor(x => x)
+                .Must(ReceiveOnlyBeforeOrAfter)
+                .WithName("Before/After")
+                .WithMessage("Please provide either Before or After or none of them");
+            RuleFor(x => x.TargetId).NotEmpty().NotNull();
+        }
+        private bool ReceiveOnlyBeforeOrAfter(PaginatedClaimRequest request)
```
Comment (andrewclowes) [inline DocumentsApi/V1/Validators/PaginatedClaimRequestValidator.cs:16]: Missing new line.

File: DocumentsApi/V1/Boundary/Response/PaginatedClaimResponse.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    public class PaginatedClaimResponse
+    {
+        public List<ClaimResponse> Claims { get; set; }
+        public Paging Paging { get; set; }
+    }
+
+    public class Paging
+    {
+        public Cursors Cursors { get; set; } = null;
+        public bool HasPreviousPage { get; set; }
```
Comment (andrewclowes) [inline DocumentsApi/V1/Boundary/Response/PaginatedClaimResponse.cs:15]: I'm not overly precious about this but are we happy with the names of these booleans? What about `hasBefore` and `hasAfter` to align with the cursors?

File: DocumentsApi/V1/UseCase/GetClaimsByTargetIdUseCase.cs [hunk 4]
```diff
@@ -1,35 +1,98 @@
...
+
+                if (claims.Count == request.Limit + 1)
+                {
+                    hasNextPage = true;
+                    claims.RemoveAt(claims.Count - 1);
+                }
+            }
+            else
+            {
+                var parsedBefore = Base64UrlHelpers.DecodeFromBase64Url(request.Before);
```
Comment (andrewclowes) [inline DocumentsApi/V1/UseCase/GetClaimsByTargetIdUseCase.cs:67]: If the decoding fails due to the endpoint being passed an encoding string which isn't Base64URL encoded then do we want to send them a bad request response to tell them there token is invalid?

File: DocumentsApi/V1/Boundary/Response/PaginatedClaimResponse.cs (hunk 3 above)
Comment (bogdan-zaharia) [reply DocumentsApi/V1/Boundary/Response/PaginatedClaimResponse.cs]: Addressed

File: DocumentsApi.Tests/V1/Helpers/Base64UrlHelpersTests.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        [Test]
+        public void DecodeReturnsExpectedString()
+        {
+            var s = "ewogICJpZCI6ICJ0aGlzLWlzLW15LWlkIgp9";
+            var result = Base64UrlHelpers.DecodeFromBase64Url(s);
+            result.Should().BeEquivalentTo(JObject.Parse("{\"id\":\"" + $"this-is-my-id" + "\"}"));
+        }
+        [Test]
```
[the remark this answers is not in the document]
Comment (bogdan-zaharia) [reply DocumentsApi.Tests/V1/Helpers/Base64UrlHelpersTests.cs:36]: Addressed

File: DocumentsApi/V1/Validators/PaginatedClaimRequestValidator.cs (hunk 2 above)
Comment (bogdan-zaharia) [reply DocumentsApi/V1/Validators/PaginatedClaimRequestValidator.cs:17]: Addressed

File: DocumentsApi/V1/UseCase/GetClaimsByTargetIdUseCase.cs (hunk 4 above)
Comment (bogdan-zaharia) [reply DocumentsApi/V1/UseCase/GetClaimsByTargetIdUseCase.cs:75]: Addressed

File: DocumentsApi/V1/UseCase/GetClaimsByTargetIdUseCase.cs [hunk 6]
```diff
@@ -1,35 +1,114 @@
...
+                    claims.RemoveAt(claims.Count - 1);
+                }
+            }
 
-            var result = new Dictionary<string, List<ClaimResponse>>()
+            else if (request.After != null)
             {
-                {"claims", new List<ClaimResponse>()}
-            };
+                Guid decodedNextPageCursorId;
```
Comment (andrewclowes) [inline DocumentsApi/V1/UseCase/GetClaimsByTargetIdUseCase.cs:54]: There's a fair amount of repeated logic in how the `before` and `after` tokens are handled. Do you think it's worth finding an abstraction for this code?

Comment (bogdan-zaharia) [reply DocumentsApi/V1/UseCase/GetClaimsByTargetIdUseCase.cs]: Addressed

**Review by andrewclowes — APPROVED** [review]

**Pull request merged**